### PR TITLE
ROC-5504 refactor full defence dashboard

### PR DESCRIPTION
**Pull request by sudheer-ajax**
### JIRA link

https://tools.hmcts.net/jira/browse/ROC-5504

### Change description

Introduced new state machine to find the different states between full defence journeys, and update the status of dashboard based on the current state where the claim ends. This is an initial draft.

### Work checklist

- [ ] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [ ] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/main/features/dashboard/routes/index.ts [hunk 1]
```diff
@@ -20,9 +22,32 @@ export default express.Router()
...
+      claimantState.findState(claimantState)
+
+      if (claimantState.is('full-defence')) {
+        claimantState = FullDefenceTransitions(eachClaim)
+        claimantState.findState(claimantState)
+      }
+
+      eachClaim.template = claimantState.getTemplate('defendant')
+    })
+
```
Comment (KiranVarma02) [inline src/main/features/dashboard/routes/index.ts]: Could you maybe extract above 2 forEach into 1 function?

File: src/main/features/dashboard/claims-state-machine/initial-transitions.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import * as StateMachine from '@taoqf/javascript-state-machine'
+import { Claim } from 'claims/models/claim'
+
+import { ResponseType } from 'claims/models/response/responseType'
+import { isPastDeadline } from 'claims/isPastDeadline'
+import { MomentFactory } from 'shared/momentFactory'
+import * as _ from 'lodash'
+import * as path from 'path'
+
+export function InitialTransitions (claim: Claim) {
```
Comment (KiranVarma02) [inline src/main/features/dashboard/claims-state-machine/initial-transitions.ts]: please add return type as `StateMachine`

File: src/main/features/dashboard/views/status/defendant/init.njk [hunk 3]
```diff
@@ -0,0 +1 @@
+Awaiting Response
```
Comment (KiranVarma02) [inline src/main/features/dashboard/views/status/defendant/init.njk]: should this be in translate too?

File: src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import * as StateMachine from '@taoqf/javascript-state-machine'
+import { Claim } from 'claims/models/claim'
+
+import { FreeMediationOption } from 'forms/models/freeMediation'
+import { DefenceType } from 'claims/models/response/defenceType'
+import { ClaimantResponseType } from 'claims/models/claimant-response/claimantResponseType'
+import * as _ from 'lodash'
+import * as path from 'path'
+
+export function FullDefenceTransitions (claim: Claim) {
```
Comment (KiranVarma02) [inline src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts]: please add function return type as `StateMachine`

File: src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      { name : 'checkSettlementOfferAcceptWithMediation', from: ['full-defence', 'fd-settlement-offer-with-mediation'], to: 'fd-made-agreement-with-mediation' },
+      { name : 'checkSettlementOfferAcceptWithoutMediation', from: ['full-defence', 'fd-settlement-offer-without-mediation'], to: 'fd-made-agreement-without-mediation' },
+
+      { name : 'checkSettledByAgreement', from: ['full-defence', 'fd-settlement-offer-with-mediation','fd-settlement-offer-without-mediation', 'fd-made-agreement-with-mediation','fd-made-agreement-without-mediation'], to: 'fd-settled-with-agreement' },
+      { name : 'checkSettledByAgreement', from: ['full-defence', 'fd-settlement-offer-reject-without-mediation'], to: 'fd-settled' }
+    ],
+    methods: {
+
+      onInvalidTransition (transition: string, from: string, to: string): void {
+        // When the transaction is not allowed then state is going to remain same
```
Comment (KiranVarma02) [inline src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts]: This is good. Just out of interest, did you see this happen anywhere? what should we do if it happens? how are we planning to handle it?

File: src/main/features/dashboard/routes/index.ts [hunk 6]
```diff
@@ -20,9 +22,32 @@ export default express.Router()
...
     const claimDraftSaved: boolean = claimDraft.document && claimDraft.id !== 0
     const responseDraftSaved = responseDraft && responseDraft.document && responseDraft.id !== 0
-
     const claimsAsDefendant: Claim[] = await claimStoreClient.retrieveByDefendantId(user)
 
+    claimsAsClaimant.forEach(function (eachClaim) {
+      let claimantState = InitialTransitions(eachClaim)
+      claimantState.findState(claimantState)
+
+      if (claimantState.is('full-defence')) {
```
Comment (robertparkinson) [inline src/main/features/dashboard/routes/index.ts]: Can you please create ENUMs for these string literals?

File: src/main/features/dashboard/routes/index.ts [hunk 7]
```diff
@@ -20,9 +22,32 @@ export default express.Router()
...
+        claimantState = FullDefenceTransitions(eachClaim)
+        claimantState.findState(claimantState)
+      }
+
+      eachClaim.template = claimantState.getTemplate('claimant')
```
Comment (robertparkinson) [inline src/main/features/dashboard/routes/index.ts]: Can you please create ENUMs for these string literals?

File: src/main/features/dashboard/routes/index.ts [hunk 8]
```diff
@@ -20,9 +22,32 @@ export default express.Router()
...
+    })
+
+    claimsAsDefendant.forEach(function (eachClaim) {
+      let claimantState = InitialTransitions(eachClaim)
+      claimantState.findState(claimantState)
+
+      if (claimantState.is('full-defence')) {
```
Comment (robertparkinson) [inline src/main/features/dashboard/routes/index.ts]: Can you please create ENUMs for these string literals?

File: src/main/features/dashboard/routes/index.ts [hunk 9]
```diff
@@ -20,9 +22,32 @@ export default express.Router()
...
+        claimantState = FullDefenceTransitions(eachClaim)
+        claimantState.findState(claimantState)
+      }
+
+      eachClaim.template = claimantState.getTemplate('defendant')
```
Comment (robertparkinson) [inline src/main/features/dashboard/routes/index.ts]: Can you please create ENUMs for these string literals?

File: src/main/app/claims/models/response/fullAdmissionResponse.ts [hunk 10]
```diff
@@ -3,11 +3,15 @@ import { ResponseType } from 'claims/models/response/responseType'
 
 import { PaymentIntention } from 'claims/models/response/core/paymentIntention'
 import { StatementOfMeans } from 'claims/models/response/statement-of-means/statementOfMeans'
+import { PaymentDeclaration } from 'claims/models/paymentDeclaration'
+import { DefenceType } from 'claims/models/response/defenceType'
 
 export interface FullAdmissionResponse extends ResponseCommon {
   responseType: ResponseType.FULL_ADMISSION
+  paymentDeclaration?: PaymentDeclaration
```
Comment (robertparkinson) [inline src/main/app/claims/models/response/fullAdmissionResponse.ts]: Why is this being added here?

File: src/main/app/claims/models/response/fullAdmissionResponse.ts [hunk 11]
```diff
@@ -3,11 +3,15 @@ import { ResponseType } from 'claims/models/response/responseType'
...
   paymentIntention?: PaymentIntention
   statementOfMeans?: StatementOfMeans
+  defenceType?: DefenceType
```
Comment (robertparkinson) [inline src/main/app/claims/models/response/fullAdmissionResponse.ts]: Why is this being added here?

File: src/main/app/claims/models/response/partialAdmissionResponse.ts [hunk 12]
```diff
@@ -18,6 +19,7 @@ export interface PartialAdmissionResponse extends ResponseCommon {
   freeMediation?: YesNoOption,
   paymentIntention?: PaymentIntention
   statementOfMeans?: StatementOfMeans
+  defenceType?: DefenceType
```
Comment (robertparkinson) [inline src/main/app/claims/models/response/partialAdmissionResponse.ts]: Why is this being added here?

File: src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts (hunk 4 above)
Comment (KiranVarma02) [reply src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts]: I would change name to start with small letter as its a function...

File: src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+import { FreeMediationOption } from 'forms/models/freeMediation'
+import { DefenceType } from 'claims/models/response/defenceType'
+import { ClaimantResponseType } from 'claims/models/claimant-response/claimantResponseType'
+import * as _ from 'lodash'
+import * as path from 'path'
+import { FullDefenceStates } from 'claims/models/claim-states/full-defence-states'
+import { FullDefenceResponse } from 'claims/models/response/fullDefenceResponse'
+
+export function fullDefenceTransitions (claim: Claim): StateMachine {
```
Comment (sabahirfan) [inline src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts]: Please reformat this file to make it consistent.  There are few places with wrong spaces.

File: src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+  return new StateMachine({
+    init : FullDefenceStates.FULL_DEFENCE,
+    transitions : [
+      {
+        name : 'checkAlreadyPaid',
```
Comment (sabahirfan) [inline src/main/features/dashboard/claims-state-machine/full-defence-transitions.ts]: e.g. extra space before colon : here

File: src/main/app/claims/models/claim.ts [hunk 15]
```diff
@@ -58,6 +59,7 @@ export class Claim {
   moneyReceivedOn: Moment
   reDetermination: ReDetermination
   reDeterminationRequestedAt: Moment
+  template: Object
```
Comment (sabahirfan) [inline src/main/app/claims/models/claim.ts]: Can we please avoid using a generic Object here?

File: src/main/features/dashboard/views/status/claimant/init.njk [hunk 16]
```diff
@@ -0,0 +1 @@
+{{ t('Awaiting Response..') }}
```
Comment (sabahirfan) [inline src/main/features/dashboard/views/status/claimant/init.njk]: do we need two periods here?

File: src/main/features/dashboard/views/status/defendant/init.njk [hunk 17]
```diff
@@ -0,0 +1 @@
+{{ t('Awaiting Response..') }}
```
Comment (sabahirfan) [inline src/main/features/dashboard/views/status/defendant/init.njk]: do we need two periods here?

File: src/test/features/dashboard/claims-state-machine/initial-transitions.ts [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+      const claim: Claim = new Claim().deserialize(sampleClaimIssueObj)
+      let claimState = initialTransitions(claim)
+      claimState.findState(claimState)
+      expect(claimState.state).to.equal('no-response')
+    })
+  })
+
+  describe('given the claim with more time requested', () => {
+    it('should extract the correct state for the claim issued', () => {
+      const claim: Claim = new Claim().deserialize({ ...sampleClaimIssueObj,moreTimeRequested: true })
```
Comment (sabahirfan) [inline src/test/features/dashboard/claims-state-machine/initial-transitions.ts]: Missing space after comma here. and other places in the same file. please reformat this as well.

File: src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+              type: 'OFFER',
+              madeBy: 'DEFENDANT',
+              offer: {
+                content: 'test',
+                completionDate: MomentFactory.parse('2019-05-01')
+              }
+            }
+          ]
+        },
+        settlementReachedAt : MomentFactory.parse('2019-03-01')
```
Comment (sabahirfan) [inline src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts]: Remove extra space before colon.

File: src/main/features/dashboard/views/status/claimant/init.njk (hunk 16 above)
Comment (sudheer-ajax) [reply src/main/features/dashboard/views/status/claimant/init.njk]: This is never going to display to the customer as we and when we finish all the transitions for part admit, full admit and full defence the state will not going to be init for any claim .... this is just for testers who can see this as and when we finish the each of part admit, full admit and full defence. even though I will remove those dot's

File: src/main/features/dashboard/views/status/defendant/init.njk (hunk 17 above)
Comment (sudheer-ajax) [reply src/main/features/dashboard/views/status/defendant/init.njk]: This is never going to display to the customer as we and when we finish all the transitions for part admit, full admit and full defence the state will not going to be init for any claim .... this is just for testers who can see this as and when we finish the each of part admit, full admit and full defence. even though I will remove those dot's

File: src/main/app/claims/models/claim.ts (hunk 15 above)
Comment (sudheer-ajax) [reply src/main/app/claims/models/claim.ts]: sure

File: src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts (hunk 19 above)
Comment (sudheer-ajax) [reply src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts]: I will format the file with CMC code style

File: src/test/features/dashboard/claims-state-machine/initial-transitions.ts (hunk 18 above)
Comment (sudheer-ajax) [reply src/test/features/dashboard/claims-state-machine/initial-transitions.ts]: I will format the code with CMC code style

Comment (sudheer-ajax) [reply src/test/features/dashboard/claims-state-machine/initial-transitions.ts]: Its done ... committed changes

File: src/main/app/claims/models/claim.ts (hunk 15 above)
Comment (sudheer-ajax) [reply src/main/app/claims/models/claim.ts]: Removed the generic Object and created an interface for the template

File: src/main/features/dashboard/views/status/claimant/init.njk (hunk 16 above)
Comment (sudheer-ajax) [reply src/main/features/dashboard/views/status/claimant/init.njk]: its done, committed the latest changes

File: src/main/features/dashboard/views/status/defendant/init.njk (hunk 17 above)
Comment (sudheer-ajax) [reply src/main/features/dashboard/views/status/defendant/init.njk]: its done, committed the latest changes

File: src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts (hunk 19 above)
Comment (sudheer-ajax) [reply src/test/features/dashboard/claims-state-machine/full-defence-transitions.ts]: its done, committed the latest changes

**Pull request merged**